### PR TITLE
Do not return immutable `_id` error from `findAndModify` for upserting same `_id`

### LINKED ISSUE #2803: Error updating a document passing its same id with the $set operator
We probably need to move that validation closer to the actual update.

The original issue is below.

---

### FerretDB version

1.2.1

### Backend

PostgreSQL

### Environment

- OS: Mac OS
- Deployment: Docker using ghcr.io/ferretdb/all-in-one:1.2.1 image
- Deployment details: Docker desktop Version 4.20.0 (109717)

### What did you do?

With MongoDB it is possibile to update a document by passing the id, but this gives an error with FerretDB.

For example, here is a mongo shell session against MongoDB 6

```shell
foo [direct: primary] test> db.coll.insertOne({_id:1})
{ acknowledged: true, insertedId: 1 }
foo [direct: primary] test> db.coll.updateOne({_id:1},{$set:{_id:1,a:1}})
{
  acknowledged: true,
  insertedId: null,
  matchedCount: 1,
  modifiedCount: 1,
  upsertedCount: 0
}

foo [direct: primary] test> db.coll.find()
[ { _id: 1, a: 1 } ]
foo [direct: primary] test> db.coll.updateOne({_id:1},{$set:{_id:2,a:1}})
MongoServerError: Performing an update on the path '_id' would modify the immutable field '_id'
foo [direct: primary] test> db.coll.updateOne({_id:1},{$set:{_id:1,a:1}})
{
  acknowledged: true,
  insertedId: null,
  matchedCount: 1,
  modifiedCount: 0,
  upsertedCount: 0
}
```

MongoDB raises an error if one tries to update the immutable _id, but it allows one to execute the script if the _id is the same.

With FerretDB the behavior is different. As you can see, FerretDB raises an error even when the _id is the same:

```shell
test> db.test.insertOne({_id:1})
{ acknowledged: true, insertedId: 1 }
test> db.test.updateOne({_id:1},{$set:{_id:1,a:1}})
MongoServerError: Performing an update on the path '_id' would modify the immutable field '_id'
```

### What did you expect to see?

We expect no error when updating a document using a command like `db.coll.updateOne({_id:1},{$set:{_id:1,a:1}})` 

### What did you see instead?

The following error is raised: 

```
MongoServerError: Performing an update on the path '_id' would modify the immutable field '_id'
```

### PR DESCRIPTION
**Pull request by chilagrow**
# Description

Closes #2803.

## Readiness checklist

- [ ] I added/updated unit tests (and they pass).
- [x] I added/updated integration/compatibility tests (and they pass).
- [ ] I added/updated comments and checked rendering.
- [ ] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

**Review by noisersup — APPROVED** [review]
LGTM

File: integration/findandmodify_compat_test.go [hunk 1]
```diff
@@ -585,9 +585,8 @@ func TestFindAndModifyCompatUpsert(t *testing.T) {
 			command: bson.D{
 				{"query", bson.D{{"_id", bson.D{{"$exists", true}}}}},
 				{"upsert", true},
-				{"update", bson.D{{"_id", "replaced"}, {"v", "replaced"}}},
+				{"update", bson.D{{"_id", "int32"}, {"v", "replaced"}}},
```
Comment (rumyantseva) [inline integration/findandmodify_compat_test.go:588]: Hmm, to me this is a bit separate test case. In the previous test case ("replace") we were trying to set the value that didn't exist in any collection, just a new _id. In this test case, we try to set the value int32 which exists at least in some collections... I'd keep both of the test cases.

Comment (chilagrow) [reply integration/findandmodify_compat_test.go:588]: I can keep both :+1:

**Review by rumyantseva — APPROVED** [review]
LGTM

**Review by noisersup — APPROVED** [review]

**Pull request merged**